localstack/services/sns/publisher.py

```diff
@@ -849,33 +882,40 @@ def _evaluate_filter_policy_conditions(
             values = ast.literal_eval(val)
             for value in values:
                 for condition in conditions:
-                    if self._evaluate_condition(value, condition, message_attributes, criteria):
+                    if self._evaluate_condition(value, condition, field_exists):
                         return True
         else:
             for condition in conditions:
                 value = val or None
-                if self._evaluate_condition(value, condition, message_attributes, criteria):
+                if self._evaluate_condition(value, condition, field_exists):
                     return True
 
         return False
 
-    def _evaluate_condition(self, value, condition, message_attributes, criteria):
+    def _evaluate_filter_policy_conditions_on_field(self, conditions, value, field_exists: bool):
+        for condition in conditions:
+            if self._evaluate_condition(value, condition, field_exists):
+                return True
+
+        return False
+
+    def _evaluate_condition(self, value, condition, field_exists: bool):
         if type(condition) is not dict:
             return value == condition
-        elif condition.get("exists") is not None:
-            return self._evaluate_exists_condition(
-                condition.get("exists"), message_attributes, criteria
-            )
+        elif (must_exist := condition.get("exists")) is not None:
+            # if must_exists is True then field_exists must be True
+            # if must_exists is False then fields_exists must be False
+            # we can use the XNOR operator
+            return not must_exist ^ field_exists
```

Comment:
Maybe we can write `(must_exists and field_exists) or (not must_exists and not field_exists)` or `(must_exists and not field_exists) and (not must_exists or not field_exists)`, not sure if clearer!

Reply:
This thoroughly confused me :sweat_smile: 
So
```
            # if must_exists is True then field_exists must be True
            # if must_exists is False then fields_exists must be False
            # we can use the XNOR operator
```
This indeed describes XNOR (I had to look that one up), which is just equivalence, so ==.
Then you proceed to use the XOR operator.
In your comment you then once describe again XNOR `(must_exists and field_exists) or (not must_exists and not field_exists)`
and then an operation that is not clear to me `(must_exists and not field_exists) and (not must_exists or not field_exists)` (first parentheses state must_exists must be true and field_exists must be false, second states one of them must be false, which is redundant at this point if must_exists must be true).
Please clarify :smile:

Reply:
Oh my, this is embarrassing. Yes, I used the XOR with a not to make XNOR, which is... `==` as you just pointed it out. And the comment the first `and` is actually a `or`. To conclude, everything here is wrong 😂 thanks a lot really! I'll just correct it to the simplest which is a simple equivalence...